contracts/adapters/OptimisticAuctionRebalanceExtension.sol

```diff
@@ -0,0 +1,390 @@
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    SPDX-License-Identifier: Apache License, Version 2.0
+*/
+
+pragma solidity 0.6.10;
+pragma experimental "ABIEncoderV2";
+
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from  "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+
+import { AddressArrayUtils } from "../lib/AddressArrayUtils.sol";
+
+import { IAuctionRebalanceModuleV1 } from "../interfaces/IAuctionRebalanceModuleV1.sol";
+import { IBaseManager } from "../interfaces/IBaseManager.sol";
+import { ISetToken } from "../interfaces/ISetToken.sol";
+import {AuctionRebalanceExtension} from "./AuctionRebalanceExtension.sol";
+import {AncillaryData } from "../lib/AncillaryData.sol";
+import { AssetAllowList } from "../lib/AssetAllowList.sol";
+import {OptimisticOracleV3Interface} from "../interfaces/OptimisticOracleV3Interface.sol";
+
+/**
+ * @title BaseOptimisticAuctionRebalanceExtension
+ * @author Index Coop
+ *
+ * @dev The contract extends `BaseAuctionRebalanceExtension` by adding an optimistic oracle mechanism for validating rules on the proposing and executing of rebalances. 
+ * It allows setting product-specific parameters for optimistic rebalancing and includes callback functions for resolved or disputed assertions.
+ */
+contract OptimisticAuctionRebalanceExtension is  AuctionRebalanceExtension, AssetAllowList {
+    using AddressArrayUtils for address[];
+    using SafeERC20 for IERC20;
+
+    /* ============ Events ============ */
+
+    event ProductSettingsUpdated(
+        IERC20 indexed setToken,
+        address indexed manager,
+        OptimisticRebalanceParams optimisticParams,
+        bytes32 indexed rulesHash
+    );
+    event RebalanceProposed(
+        ISetToken indexed setToken,
+        IERC20 indexed quoteAsset,
+        address[] oldComponents,
+        address[] newComponents,
+        AuctionExecutionParams[] newComponentsAuctionParams,
+        AuctionExecutionParams[] oldComponentsAuctionParams,
+        bool shouldLockSetToken,
+        uint256 rebalanceDuration,
+        uint256 positionMultiplier
+    );
+
+    event AssertedClaim(
+        IERC20 indexed setToken,
+        address indexed _assertedBy,
+        bytes32 indexed rulesHash,
+        bytes32 _assertionId,
+        bytes _claimData
+    );
+
+    event ProposalDeleted(
+        bytes32 assertionID, 
+        Proposal indexed proposal
+        );
+
+    /* ============ Structs ============ */
+
+        struct AuctionExtensionParams {
+            IBaseManager baseManager;     // Manager Contract of the set token for which to deploy this extension
+            IAuctionRebalanceModuleV1 auctionModule; // Contract that rebalances index sets via single-asset auctions
+            bool useAssetAllowlist;     // Bool indicating whether to use asset allow list
+            address[] allowedAssets;    // Array of allowed assets
```

Comment:
are we storing these twice with AssetAllowList inheritance?

Reply:
Nope no deuplications, this is just a struct definition that we use to encode the arguments to be sent to the constructor.